pkg/addonfactory/chartaddonfactory/chartaddonagent.go

```diff
@@ -0,0 +1,132 @@
+const annotationValuesName string = "addon.open-cluster-management.io/values"
+
+// the reserved fields for template config
+const (
+	clusterName           string = "clusterName"
+	addonInstallNamespace string = "addonInstallNamespace"
+	hubKubeConfigSecret   string = "hubKubeConfigSecret"
+)
+
+type ChartAgentAddon struct {
+	decoder           runtime.Decoder
+	chart             *chart.Chart
+	values            interface{}
+	agentAddonOptions agent.AgentAddonOptions
+}
+
+func newChartAgentAddon(
+	scheme *runtime.Scheme,
+	chart *chart.Chart,
+	values interface{},
+	agentAddonOptions agent.AgentAddonOptions) *ChartAgentAddon {
+	return &ChartAgentAddon{
+		decoder:           serializer.NewCodecFactory(scheme).UniversalDeserializer(),
+		chart:             chart,
+		values:            values,
+		agentAddonOptions: agentAddonOptions}
+}
+
+func (a *ChartAgentAddon) Manifests(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	var objects []runtime.Object
+
+	values, err := a.getValues(cluster, addon)
+	if err != nil {
+		return objects, err
+	}
+
+	helmEngine := engine.Engine{
+		Strict:   true,
+		LintMode: false,
+	}
+
+	templates, err := helmEngine.Render(a.chart, values)
+	if err != nil {
+		return objects, err
+	}
+	for _, data := range templates {
+		if len(data) == 0 {
+			continue
+		}
+
+		object, _, err := a.decoder.Decode([]byte(data), nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, object)
+	}
+
+	return objects, nil
+}
+
+func (a *ChartAgentAddon) GetAgentAddonOptions() agent.AgentAddonOptions {
+	return a.agentAddonOptions
+}
+
+func (a *ChartAgentAddon) getValues(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (map[string]interface{}, error) {
+	overrideValues := structs.Map(a.values)
+	overrideValues[clusterName] = cluster.GetName()
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	overrideValues[addonInstallNamespace] = installNamespace
+	overrideValues[hubKubeConfigSecret] = fmt.Sprintf("%s-hub-kubeconfig", a.agentAddonOptions.AddonName)
+
+	// users can override the template config fields by the annotation of addon cr.
+	// the key of the annotation is `templateConfig`, the value is a json string which has the config struct.
+	// for example: "addon.open-cluster-management.io/values": `{"NodeSelector":{"host":"ssd"},"Image":"quay.io/helloworld:2.4"}`
+	annotations := addon.GetAnnotations()
+	if len(annotations[annotationValuesName]) != 0 {
+		annotationValues := map[string]interface{}{}
+		err := json.Unmarshal([]byte(annotations[annotationValuesName]), &annotationValues)
+		if err != nil {
+			klog.Error("failed to unmarshal the values from annotation of addon cr %v. err:%v", annotationValues, err)
+			return nil, err
+		}
+
+		for k, v := range annotationValues {
+			overrideValues[k] = v
```

Comment:
this is suspicious, what if user define addonInstallNamespace in the values. Which should be overridden

Reply:
I define `Values map[string]interface{}` to replace the `Values`  defined in chart lib.
user needs to define  ```
getValuesFunc func(cluster *clusterv1.ManagedCluster,
	addon *addonapiv1alpha1.ManagedClusterAddOn) (Values, error). 
```
and can use `GetValuesFromAddonAnnotation` to get values in annotation of addon CR, and use `MergeValues` to merge 2 Values.